torch/ao/quantization/fx/_model_report/model_report.py

```diff
@@ -0,0 +1,105 @@
+from typing import Dict, Set, Tuple
+
+import torch
+from torch.ao.quantization.fx._model_report.detector import DetectorBase
+from torch.ao.quantization.fx.graph_module import GraphModule
+
+
+DEFAULT_MODEL_REPORT_OBSERVER_PRE_EXTENSION = ".model_report_pre_observer"
+DEFAULT_MODEL_REPORT_OBSERVER_POST_EXTENSION = ".model_report_post_observer"
+
+
+class ModelReport:
+    r"""
+    Generates report and collects statistics
+        Used to provide users suggestions on possible model configuration improvements
+
+    Currently supports generating reports on:
+    - Suggestions for dynamic vs static quantization for linear layers (Graph Modules)
+
+    * :attr:`desired_report_detectors` The set of Detectors representing desired reports from the ModelReport class
+        Make sure that these are all unique types of detectors [do not have more than 1 of the same class]
+
+    Proper Use:
+    1.) Initialize ModelReport object with reports of interest by passing in initialized detector objects
+    2.) Prepare your model with prepare_fx
+    3.) Call model_report.prepare_detailed_calibration on your model to add relavent observers
+    4.) Callibrate your model with data
+    5.) Call model_report.generate_report on your model to generate report and optionally remove added observers
+
+    """
+
+    def __init__(self, desired_report_detectors: Set[DetectorBase]):
+
+        if len(desired_report_detectors) == 0:
+            raise ValueError("Should include at least 1 desired report")
+
+        # keep the reports private so they can't be modified
+        self.__desired_report_detectors = desired_report_detectors
+        self.__desired_reports = set([detector.get_detector_name() for detector in desired_report_detectors])
+
+        # keep a mapping of desired reports to observers of interest
+        # this is to get the readings, and to remove them, can create a large set
+        # this set can then be used to traverse the graph and remove added observers
+        self.__report_to_observers_of_interest: Dict[str, Set[str]] = {}
```

Comment:
What is the value here? Is it the FQNs of the observers? (If so maybe we should include that in the name of this field?)

Reply:
Yeah it maps report name to observer fqns relevant to that report. I updated it to better reflect.